src/main/java/com/google/sps/servlets/MatchServlet.java

```diff
@@ -21,48 +25,74 @@
+    
+    userService = UserServiceFactory.getUserService();
+    com.google.appengine.api.users.User currentGoogleUser = userService.getCurrentUser();
+    if(currentGoogleUser == null) {
+      currentUser = null;
+    } else {
+      currentUser = userRepository.getUser(currentGoogleUser);
+    }
 
-    //Set MatchRepository and Current User as ServletContext so that they can be accessed by all servlets
-    getServletContext().setAttribute("matchRepository", repository);
-    getServletContext().setAttribute("currentUser", testUser);
+    getServletContext().setAttribute("currentUser", currentUser);
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    setLoggedInUser();
+    Gson gson = new Gson();
     ServletContext servletContext = getServletContext();
-    MatchRepository testRepository = (MatchRepository) servletContext.getAttribute("matchRepository");
-    User testUser = (User) servletContext.getAttribute("currentUser");
+    User currentUser = (User) servletContext.getAttribute("currentUser");
 
-    Collection<User> matches = testRepository.getMatchesForUser(testUser);
+    //is user logged in?
+    if(currentUser == null) {
+        //return null for matches so that page redirects to logged out homepage
+        response.getWriter().println(gson.toJson(null));
+    } else {
+        MatchRepository matchRepository = (MatchRepository) servletContext.getAttribute("matchRepository");
+        Collection<User> matches = matchRepository.getMatchesForUser(currentUser.getId());
 
-    Gson gson = new Gson();
-    response.setContentType("application/json;");
-    response.getWriter().println(gson.toJson(matches));
+        response.setContentType("application/json;");
+        response.getWriter().println(gson.toJson(matches));
+    }
   }
 
   @Override
   public synchronized void doPost(HttpServletRequest request, HttpServletResponse response)
       throws IOException {
     ServletContext servletContext = getServletContext();
-    MatchRepository testRepository = (MatchRepository) servletContext.getAttribute("matchRepository");
-    User testUser = (User) servletContext.getAttribute("currentUser");
+    MatchRepository matchRepository = (MatchRepository) servletContext.getAttribute("matchRepository");
+    User currentUser = (User) servletContext.getAttribute("currentUser");
 
     if(request.getParameterValues("new-matches") != null) {
       String[] matchesToSave = request.getParameterValues("new-matches");
 
       for (String matchName : matchesToSave) {
         User newMatch = new Gson().fromJson(matchName, User.class);
-        testRepository.addMatch(testUser, newMatch);
+        matchRepository.addMatch(currentUser.getId(), newMatch);
       }
     } else {
       System.err.println("new-matches is null in MatchServlet doPost()");
     }
+
+    //update global matchRepo
+    servletContext.setAttribute("matchRepository", matchRepository);
```

Comment:
Is setting the attribute after changing it necessary for servletContexts?

Reply:
I tried taking this out and it still works, so I don't think it was necessary. Thanks!

Reply:
I figured. For things in datastore this pattern is needed since it's serializing/deserializing, but the servletContext attribute map is likely just a collection, or something similar, so the value you get from there won't change so long as someone doesn't overwrite the matchRepository with another instance.